ext/mysqlnd/mysqlnd_wireprotocol.c

```diff
@@ -2487,9 +2495,14 @@ MYSQLND_METHOD(mysqlnd_protocol, send_command_handle_OK)(
 
 	payload_decoder_factory->m.init_ok_packet(&ok_response);
 	DBG_ENTER("mysqlnd_protocol::send_command_handle_OK");
+
 	if (FAIL == (ret = PACKET_READ(payload_decoder_factory->conn, &ok_response))) {
-		DBG_INF("Error while reading OK packet");
-		SET_CLIENT_ERROR(error_info, CR_MALFORMED_PACKET, UNKNOWN_SQLSTATE, "Malformed packet");
+		if (command == COM_PING && error_info->error_no == CR_SERVER_GONE_ERROR) {
+			SET_EMPTY_ERROR(error_info);
+		} else {
+			DBG_INF("Error while reading OK packet");
+			SET_CLIENT_ERROR(error_info, CR_MALFORMED_PACKET, UNKNOWN_SQLSTATE, "Malformed packet");
```

Comment:
Personally, I feel that if you get a gone away error, you should use gone away instead of malformed.